rustc_span: Optimize span parent get/set methods

petrochenkov opened the pull request:
Like https://github.com/rust-lang/rust/pull/125017, but for span parents.

r? @cjgillot

petrochenkov committed:
rustc_span: Optimize span parent get/set methods

petrochenkov commented on compiler/rustc_span/src/span_encoding.rs:
```diff
@@ -401,6 +401,63 @@ impl Span {
+        };
+        if let Some(parent) = parent {
+            // FIXME: Is this tracking necessary?
+            (*SPAN_TRACK)(parent);
```
@cjgillot 
Is it necessary to track the span of the parent when we are just getting that parent's id?
From what I understand we don't need it because we are not getting an actual absolute span position (like we'd do in e.g. `span.lo()`), but I'm not entirely certain.

cjgillot replied:
This tracking is unnecessary, I agree.

petrochenkov commented on compiler/rustc_span/src/span_encoding.rs:
```diff
@@ -401,6 +401,63 @@ impl Span {
+        // We could not keep the span in the same inline format, fall back to the complete logic.
+        if let Some(old_parent) = data.parent {
+            // FIXME: Is this tracking necessary?
+            (*SPAN_TRACK)(old_parent);
```
I've reproduced the pre-existing logic precisely and it tracks the old span.
It's pretty counter intuitive, but it may indeed be correct? We are effectively taking the old span's absolute position (which needs tracking) and produce a new span with a new parent out of it.
Similarly to https://github.com/rust-lang/rust/pull/126544/files#r1641782800 I'm also not sure.

cjgillot replied:
Yes, we should track the old span. When the old span has a parent, it encodes the absolute lo/hi byte positions, but incr. comp. is convinced it only encodes positions relative to the parent. If we output a span with no parent, incr. comp. will see those absolute positions appear out of thin air.
So we need to tell incr. comp. that we are accessing the absolute positions with `SPAN_TRACK`. The only case when it's not required is when `parent == old_parent`.

petrochenkov commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 90b72b581bbc6c601c8e6f77bc7a28ff89098ae0 with merge eab5a5bd74528e95727f5e553f1d0fe06bcac85e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"90b72b581bbc6c601c8e6f77bc7a28ff89098ae0","merge_sha":"eab5a5bd74528e95727f5e553f1d0fe06bcac85e"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9535472203/job/26282521294)
Build commit: eab5a5bd74528e95727f5e553f1d0fe06bcac85e (`eab5a5bd74528e95727f5e553f1d0fe06bcac85e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9535472203/job/26282521294"},"merge_sha":"eab5a5bd74528e95727f5e553f1d0fe06bcac85e"} -->

rust-timer commented:
Queued eab5a5bd74528e95727f5e553f1d0fe06bcac85e with parent 12b33d36f3de0fb30ca08c49f5b2f4fa111861a4, future [comparison URL](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=eab5a5bd74528e95727f5e553f1d0fe06bcac85e).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (eab5a5bd74528e95727f5e553f1d0fe06bcac85e): [comparison URL](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=eab5a5bd74528e95727f5e553f1d0fe06bcac85e&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=eab5a5bd74528e95727f5e553f1d0fe06bcac85e&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.3%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.2%, 0.4%]   | 8     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.4%, -0.2%] | 6     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -0.1% | [-0.4%, 0.3%]  | 8     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=eab5a5bd74528e95727f5e553f1d0fe06bcac85e&stat=max-rss)
<details>
<summary>Results (primary -2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.6% | [-2.6%, -2.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=eab5a5bd74528e95727f5e553f1d0fe06bcac85e&stat=cycles:u)
<details>
<summary>Results (primary 2.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.0% | [2.0%, 2.0%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.0% | [2.0%, 2.0%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=eab5a5bd74528e95727f5e553f1d0fe06bcac85e&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 672.776s -> 671.266s (-0.22%)
**Artifact size**: 319.83 MiB -> 320.50 MiB (0.21%)

petrochenkov commented:
Statistics (similar to https://github.com/rust-lang/rust/pull/125017#issuecomment-2150619096).

```
// from standard library
// apparently incremental is not enabled for stdlib despite `rust.incremental = true` in the config
91546 inline_ctxt: None -> None
   41 partially_interned: None -> None

// from ui test suite
// incremental is typically disabled
364420 inline_ctxt: None -> None
  3957 inline_ctxt: None -> Some
   188 inline_parent: Some -> Some
    93 partially_interned: Some -> Some
    13 partially_interned: None -> None
    11 inline_parent: Some -> None
     2 partially_interned: Some -> None

// from incremental test suite
107656 inline_ctxt: None -> Some
  8983 inline_ctxt: None -> None
  3642 inline_parent: Some -> Some
  2789 partially_interned: Some -> Some
    90 partially_interned: Some -> None
    42 inline_parent: Some -> None
     5 partially_interned: None -> Some
```

petrochenkov commented:
Updated in accordance with https://github.com/rust-lang/rust/pull/126544#issuecomment-2173187242.
I also added some comment updates for `map_ctxt` missing from the accidentally merged https://github.com/rust-lang/rust/pull/126591.
The questions about dependency tracking are still relevant though.

cjgillot commented:
r=me with the points about tracking resolved.

petrochenkov commented:
@bors r=cjgillot

bors commented:
:pushpin: Commit 4d3b617911f5903c5b4e6189fa3dea99734e1bc5 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 4d3b617911f5903c5b4e6189fa3dea99734e1bc5 -->
<!-- homu: {"type":"Approved","sha":"4d3b617911f5903c5b4e6189fa3dea99734e1bc5","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 4d3b617911f5903c5b4e6189fa3dea99734e1bc5 with merge a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd...
<!-- homu: {"type":"BuildStarted","head_sha":"4d3b617911f5903c5b4e6189fa3dea99734e1bc5","merge_sha":"a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9605812215/job/26497346176)
Approved by: cjgillot
Pushing a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9605812215/job/26497346176"},"merge_sha":"a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd"} -->

rust-timer commented:
Finished benchmarking commit (a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd): [comparison URL](https://perf.rust-lang.org/compare.html?start=684b3553f70148ded97a80371c2405984d4f6aa7&end=a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=684b3553f70148ded97a80371c2405984d4f6aa7&end=a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.7% | [0.6%, 0.8%] | 4     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=684b3553f70148ded97a80371c2405984d4f6aa7&end=a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=684b3553f70148ded97a80371c2405984d4f6aa7&end=a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd&stat=cycles:u)
<details>
<summary>Results (secondary 0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [2.8%, 2.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=684b3553f70148ded97a80371c2405984d4f6aa7&end=a9c8887c7d548abc6c3e87f7d6fa02a0e95880bd&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 700.231s -> 701.948s (0.25%)
**Artifact size**: 324.89 MiB -> 325.02 MiB (0.04%)
